### ISSUE TITLE
SignalR in uwp cannot connect to hub

### ISSUE THREAD

**Issue by Jerry19968** [SignalR/SignalR#3705]
just a brief introduction to the project I am working on. Firstly, this project is on IOT, raspberry pi 2 model B and some sensors. For this case, I am using a button switch to be checked on every seconds. Then it will write out 0 or 1 if the button is being triggered in te output console. So, I want to send data to my cloud db server via signalR which is using a hub and the hub methods.

Problem: The connection between my client(UWP) cannot connect to the hub which is published on the internet(/signalR).

I have tried a the methods that I found on the internet but it is not working for me. Method: Replace the Microsoft.AspNet.SignalR.Client.dll in the folder and add reference to the downloaded one. 

As for the Hub, I store it in a web api and published to azure web service.

An exception of type 'System.InvalidOperationException' occurred in Microsoft.AspNet.SignalR.Client.dll but was not handled in user code < Error

Microsoft.AspNet.SignalR.Client.ConnectionState.Connecting < cannot complete the connection

    System.InvalidOperationException was unhandled by user code
      HResult=-2146233079
      Message=The connection has not been established.
      Source=Microsoft.AspNet.SignalR.Client
      StackTrace:
           at Microsoft.AspNet.SignalR.Client.Connection.Send(String data)
           at Microsoft.AspNet.SignalR.Client.Hubs.HubProxy.Invoke[TResult,TProgress](String method, Action`1 onProgress, Object[] args)
           at Microsoft.AspNet.SignalR.Client.Hubs.HubProxy.Invoke(String method, Object[] args)
           at Button.signalR.addData(Int32 i)
           at Button.MainPage.Timer_Tick(Object sender, Object e)
      InnerException: 

^ Error

**My Hub**

    using System;
    using System.Collections.Generic;
    using System.Linq;
    using System.Web;
    using Microsoft.AspNet.SignalR;
    using System.Data.Entity;
    using WebApplication5.Models;

    namespace WebApplication5
    {
    public class queryHub : Hub
    {
        public void AddData(bool i)
        {
            try
            {
                using (var db = new db2Entities())
                {
                    var b = db.Tests.Create();
                    b.isOn = i;
                    db.Tests.Add(b);
                    db.SaveChanges();
                };
                
            }
            catch (Exception ex)
            {
                Clients.Caller.reportError("Unable to create bug.");
            }
        }
    }
    }

**My Startup class**

    using System;
    using System.Collections.Generic;
    using System.Linq;
    using System.Web;
    using Microsoft.Owin;
    using Owin;
    using Microsoft.AspNet.SignalR;

    [assembly: OwinStartup(typeof(WebApplication5.Startup))]

    namespace WebApplication5
    {
    public class Startup
    {
        public void Configuration(IAppBuilder app)
        {
            // Any connection or hub wire up and configuration should go here
            app.MapSignalR("/signalr", new HubConfiguration());
        }

    }
    }

**UWP signalR.cs**

    using System;
    using System.Collections.Generic;
    using System.Linq;
    using System.Text;
    using System.Threading.Tasks;
    using Microsoft.AspNet.SignalR;
    using Windows.UI.Core;
    using Microsoft.AspNet.SignalR.Client;
    using System.ComponentModel;

    namespace Button
    {
    class signalR : INotifyPropertyChanged
    {
        private HubConnection _hubConnection;
        private IHubProxy _proxy;
        private CoreDispatcher _dispatcher;
        public signalR()
        {
            _dispatcher = CoreWindow.GetForCurrentThread().Dispatcher;
        }
        public string bit { get; set; }

        public event PropertyChangedEventHandler PropertyChanged;
        public async void startConnection()
        {
            var _hubConnection = new               HubConnection("http://testwa.azurewebsites.net/signalr");
            _proxy = _hubConnection.CreateHubProxy("queryHub");
            await _hubConnection.Start();
        }

        public void addData(int i)
        {
            if(i == 0)
            {
                _proxy.Invoke("addData", false);
                startConnection();
            }
            else
            {
                _proxy.Invoke("addData", true);
                startConnection();
            }
        }

    }
    }

**MainPage.xaml.cs**

    using System;
    using System.Collections.Generic;
    using System.IO;
    using System.Linq;
    using System.Runtime.InteropServices.WindowsRuntime;
    using Windows.Foundation;
    using Windows.Foundation.Collections;
    using Windows.UI.Xaml;
    using Windows.UI.Xaml.Controls;
    using Windows.UI.Xaml.Controls.Primitives;
    using Windows.UI.Xaml.Data;
    using Windows.UI.Xaml.Input;
    using Windows.UI.Xaml.Media;
    using Windows.UI.Xaml.Navigation;
    using Microsoft.AspNet.SignalR.Client;

    // The Blank Page item template is documented at     http://go.microsoft.com/fwlink/?LinkId=402352&clcid=0x409

    using Windows.Devices.Gpio;
    using System.Threading;
    using System.Diagnostics;

    namespace Button
    {
    /// <summary>
    /// An empty page that can be used on its own or navigated to within a Frame.
    /// </summary>
    public sealed partial class MainPage : Page
    {
        signalR x = new signalR();
        private const int BUTTON_PIN = 17;

        GpioPin buttonPin;

        private DispatcherTimer timer = new DispatcherTimer();

        private int a = 0;

        private void InitGPIO()
        {
            var gpio = GpioController.GetDefault();
            if (gpio == null)
            {
                buttonPin = null;
                return;
            }

            buttonPin = gpio.OpenPin(BUTTON_PIN);

            if (buttonPin.IsDriveModeSupported(GpioPinDriveMode.InputPullUp))
                buttonPin.SetDriveMode(GpioPinDriveMode.InputPullUp);
            else
                buttonPin.SetDriveMode(GpioPinDriveMode.Input);

            //---set a time to filter one noise caused by the pressing of a button---
            buttonPin.DebounceTimeout = TimeSpan.FromMilliseconds(50);

            //---register for the ValueChanged event so the ButtonPin_ValueChanged function is called when the button is pressed---
            buttonPin.ValueChanged += ButtonPin_ValueChanged;
        }

        private void ButtonPin_ValueChanged(GpioPin sender, GpioPinValueChangedEventArgs args)
        {
            //---toggle the state of the LED every time the button is pressed---
            if (args.Edge == GpioPinEdge.FallingEdge) //---from HIGH to LOW--- 
            {
                a = (a == 0) ?
                    1 : 0;
            }
        }

        public MainPage()
        {
            this.InitializeComponent();
            InitGPIO();

            timer.Interval = TimeSpan.FromSeconds(1);
            timer.Tick += Timer_Tick;
            timer.Start();
        }

        private void Timer_Tick(object sender, object e)
        {
            Debug.Write(a);
            x.addData(a);
        }
    }
    }

**Debug Information**

[Debug 1][1]

[Debug 2][2]

  [1]: http://i.stack.imgur.com/z0quP.png
  [2]: http://i.stack.imgur.com/TJa8e.png

**Comment by moozzyk** [comment]
Looks like a dupe of https://github.com/SignalR/SignalR/issues/3547

**Comment by Jerry19968** [comment]
@moozzyk it is not a dupe, I have been to that git hub and tried the methods there. It did not work for me, that's why I posted this question. Just to add on, I am a student trying out with mini projects. 

Thank you for replying

**Comment by moozzyk** [comment]
These calls seem to be in reverse order: 
```
    _proxy.Invoke("addData", false);
    startConnection();
```

you first need to start the connection and then can send data. Note that you need to start the connection only once and then you can call invoke multiple times. You should not start a connection each time you want to send data.

**Comment by MichaelCarey** [comment]
@moozzyk I see this is closed but I'm also unable to start a connection in uwp using SignalR.Client 2.2.2. Works fine in x86 but not for ARM! Appears to have been a problem pre 2.2.1. Has this been fixed? 
Thanks!

**Comment by moozzyk** [comment]
This was fixed in 2.2.1 and I tested the fix with Raspberry PI 2 (which is arm) running Win 10 IoT Core. I did make sure the Release version worked. I don't think anything changed in 2.2.2 so it should still work. The bug used to tracking the issue was #3547.

**Comment by MichaelCarey** [comment]
@moozzyk  Thanks for reply. I just created a new new project (UWP), added the SerialR client and iot extensions for UWP. Added the following three lines of SignalR Code and that's basically it!.  

                    var hubConnection = new HubConnection("http://localhost:51240/");
                    iotHub = hubConnection.CreateHubProxy("IotHub");
                    await hubConnection.Start();

This works with x86 on my local machine (I can see the connection on server. In fact I have a lot working on the x86!). However, when I change to ARM/Remote and deploy to the Raspberry Pi (Windows iot 10.0.15063.414; .NetCore UWP 5.4.0) .... it fails. "hubConnection" won't start.   I get the following exception... I'm probably missing something obvious just can't see it .. sorry! 

System.Net.Http.HttpRequestException: An error occurred while sending the request. ---> System.Runtime.InteropServices.COMException: The text associated with this error code could not be found.

A connection with the server could not be established

   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Net.Http.HttpHandlerToFilter.<SendAsync>d__4.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Net.Http.HttpClientHandler.<SendAsync>d__86.MoveNext()
   --- End of inner exception stack trace ---
   at System.Net.Http.HttpClientHandler.<SendAsync>d__86.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Net.Http.HttpClient.<FinishSendAsync>d__58.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
   at App1.MainPage.<SetupHub>d__7.MoveNext()

**Comment by MichaelCarey** [comment]
Hi Pawel,
Thanks for looking at this. I know that you have tested this but there just
seems to be something wrong.

I have tried 2.2.1 with the same result. Perhaps I'm doing something stupid
but can't see it! I created a UWP project with two line of code:
*          var hubConnection = new HubConnection("http://localhost:51240/
<http://localhost:51240/>");*

*          await hubConnection.Start();*

*It works in x86* *but not with ARM *(Rasp Pi 3, Win iot 10.0.15063, .Net
Core.UWP 6.0.4).

I have also build from source (Microsoft.AspNet.SignalR.Client.UWP). and
get the following exception:

System.Resources.MissingManifestResourceException:
MissingManifestResource_ResWFileNotLoaded,
Microsoft.AspNet.SignalR.Client.Resources
at System.Resources.ResourceManager.GetString(String name,
Globalization.CultureInfo culture)
at Microsoft.AspNet.SignalR.Client.HubConnection.OnClosed()
at Microsoft.AspNet.SignalR.Client.Connection.Disconnect()
at Microsoft.AspNet.SignalR.Client.Connection.b__142_1()
at System.Action.Invoke()
at
Microsoft.AspNet.SignalR.TaskAsyncHelper.<>c__DisplayClass19_1.b__0(Threading.Tasks.Task
t)
--- End of stack trace from previous location where exception was thrown ---
at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
at SharedLibrary!+0x366a33
at SharedLibrary!+0x367031
at SharedLibrary!+0x367b49
at SharedLibrary!+0x367b0f
at cc200!+0x53c7db

Much appreciate any direction! Thanks!


Mike
